File: NiceReads/Backend/TOS/Common/TOS.CQRS/Dispatchers/Queries/QueryDispatcher.cs
using Microsoft.Extensions.Logging;
using System;
using System.Threading.Tasks;
using TOS.CQRS.Executions.Queries;
using TOS.CQRS.Handlers;
using TOS.CQRS.Handlers.Queries;

namespace TOS.CQRS.Dispatchers.Queries
{
    public class QueryDispatcher : IQueryDispatcher
    {
        private readonly IHandlerExecutor _handlerExecutor;
        private readonly IExecutionHandlerProvider _executionHandlerProvider;
        private readonly ILogger<QueryDispatcher> _logger;

        public QueryDispatcher(IHandlerExecutor handlerExecutor,
            IExecutionHandlerProvider executionHandlerProvider,
            ILogger<QueryDispatcher> logger)
        {
            _handlerExecutor = handlerExecutor;
            _executionHandlerProvider = executionHandlerProvider;
            _logger = logger;
        }

        public TResult Execute<TQuery, TResult>(TQuery query) where TQuery : IQuery<TResult>
        {
            IQueryHandler<TQuery, TResult> handler = _executionHandlerProvider.GetHandlerFor<IQueryHandler<TQuery, TResult>>();
            try
            {
                return _handlerExecutor.Execute<TQuery, IQueryHandler<TQuery, TResult>, TResult>(handler, query);
            }
            catch (Exception ex)
            {
                LogError<TQuery>(ex, handler);
                throw;
            }
        }

        public async Task<TResult> ExecuteAsync<TQuery, TResult>(TQuery query) where TQuery : IAsyncQuery<TResult>
        {
            IAsyncQueryHandler<TQuery, TResult> handler = _executionHandlerProvider.GetHandlerFor<IAsyncQueryHandler<TQuery, TResult>>();
            try
            {
                return await _handlerExecutor.ExecuteAsync<TQuery, IAsyncQueryHandler<TQuery, TResult>, TResult>(handler, query);
            }
            catch (Exception ex)
            {
                LogError<TQuery>(ex, handler);
                throw;
            }
        }

        private void LogError<T>(Exception ex, object handler)
        {
            _logger.LogError(ex, "Error when execution {Query} by handler {Handler}.", typeof(T).FullName, handler.GetType().FullName);
        }
    }

}

File: NiceReads/Backend/Server/tos-server/src/cqrs/dispatchers/queries/QueryDispatcher.ts
import { inject, injectable } from 'inversify';
import { Types } from '../../../types';
import { IAsyncQuery } from '../../executions/queries/async.query';
import { IQuery } from '../../executions/queries/query';
import { IHandlerExecutor } from '../../handlers/IHandlerExecutor';
import { IAsyncQueryHandler } from '../../handlers/queries/IAsyncQueryHandler';
import { IQueryHandler } from '../../handlers/queries/IQueryHandler';
import { IExecutionHandlerProvider } from '../IExecutionHandlerProvider';
import { IQueryDispatcher } from './IQueryDispatcher';

@injectable()
export class QueryDispatcher implements IQueryDispatcher {

    constructor(
        @inject(Types.EXECUTOR_HANDLER_PROVIDER) private readonly executionHandlerProvider: IExecutionHandlerProvider,
        @inject(Types.HANDLER_EXECUTOR) private readonly handlerExecutor: IHandlerExecutor
        ) { }

    execute<TQuery extends IQuery<TResult>, TResult>(query: TQuery): TResult {
        const handlerIdentifier: string = query.constructor.name;
        const handler: IQueryHandler<TQuery, TResult> =
         this.executionHandlerProvider.getHandlerFor<IQueryHandler<TQuery, TResult>>(handlerIdentifier);
        try{
            return this.handlerExecutor.executeWithResult(handler, query);
        } catch (error) {
            throw error;
        }
    }

    async executeAsync<TAsyncQuery extends IAsyncQuery<TResult>, TResult>(asyncQuery: TAsyncQuery): Promise<TResult> {
        const handlerIdentifier: string = asyncQuery.constructor.name;;
        const handler: IAsyncQueryHandler<TAsyncQuery, TResult> =
         this.executionHandlerProvider.getHandlerFor<IAsyncQueryHandler<TAsyncQuery, TResult>>(handlerIdentifier);
        try{
            return await this.handlerExecutor.executeWithResultAsync(handler, asyncQuery);
        } catch (error) {
            throw error;
        }
    }
}